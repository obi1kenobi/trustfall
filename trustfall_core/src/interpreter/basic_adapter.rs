use std::fmt::Debug;

use crate::ir::{EdgeParameters, Eid, FieldValue, Vid};

use super::{Adapter, DataContext, InterpretedQuery};

/// An iterator of vertices representing data points we are querying.
pub type VertexIterator<'vertex, VertexT> = Box<dyn Iterator<Item = VertexT> + 'vertex>;

/// An iterator of query contexts: bookkeeping structs we use to build up the query results.
///
/// Each context represents a possible result of the query. At each query processing step,
/// all the contexts at that step have fulfilled all the query conditions thus far.
///
/// This type is usually an input to adapter resolver functions. Calling those functions
/// asks them to resolve a property, edge, or type coercion for the particular vertex
/// the context is currently processing at that point in the query.
pub type ContextIterator<'vertex, VertexT> =
    Box<dyn Iterator<Item = DataContext<VertexT>> + 'vertex>;

/// Iterator of (context, outcome) tuples: the output type of most resolver functions.
///
/// Resolver functions produce an output value for each context:
/// - resolve_property() produces that property's value;
/// - resolve_neighbors() produces an iterator of neighboring vertices along an edge;
/// - resolve_coercion() gives a bool representing whether the vertex is of the desired type.
///
/// This type lets us write those output types in a slightly more readable way.
pub type ContextOutcomeIterator<'vertex, VertexT, OutcomeT> =
    Box<dyn Iterator<Item = (DataContext<VertexT>, OutcomeT)> + 'vertex>;

pub trait BasicAdapter<'vertex> {
    /// The type of vertices in the dataset this adapter queries.
    /// It's frequently a good idea to use an Rc<...> type for cheaper cloning here.
    type Vertex: Clone + Debug + 'vertex;

    /// Produce an iterator of vertices for the specified starting edge.
    ///
    /// Starting edges are ones where queries are allowed to begin.
    /// They are defined directly on the root query type of the schema.
    /// For example, `Foo` is the starting edge of the following query:
    /// ```graphql
    /// query {
    ///     Foo {
    ///         bar @output
    ///     }
    /// }
    /// ```
    ///
    /// The caller guarantees that:
    /// - The specified edge is a starting edge in the schema being queried.
    /// - Any parameters the edge requires per the schema have values provided.
    fn resolve_starting_vertices(
        &mut self,
        edge_name: &str,
        parameters: Option<&EdgeParameters>,
    ) -> VertexIterator<'vertex, Self::Vertex>;

    /// Resolve the value of a vertex property over an iterator of query contexts.
    ///
    /// Each context in the `contexts` argument has an active vertex, which is
    /// either `None`, or a `Some(Self::Vertex)` value representing a vertex
    /// of type `type_name` defined in the schema.
    ///
    /// This function resolves the property value on that active vertex.
    ///
    /// The caller guarantees that:
    /// - `type_name` is a type or interface defined in the schema.
    /// - `property_name` is either a property field on `type_name` defined in the schema,
    ///   or the special value `"__typename"` requesting the name of the vertex's type.
    /// - When the active vertex is `Some(...)`, it's a vertex of type `type_name`:
    ///   either its type is exactly `type_name`, or `type_name` is an interface that
    ///   the vertex's type implements.
    ///
    /// The returned iterator must satisfy these properties:
    /// - Produce `(context, property_value)` tuples with the property's value for that context.
    /// - Produce contexts in the same order as the input `contexts` iterator produced them.
    /// - Produce property values whose type matches the property's type defined in the schema.
    /// - When a context's active vertex is `None`, its property value is `FieldValue::Null`.
    fn resolve_property(
        &mut self,
        contexts: ContextIterator<'vertex, Self::Vertex>,
        type_name: &str,
        property_name: &str,
    ) -> ContextOutcomeIterator<'vertex, Self::Vertex, FieldValue>;

    /// Resolve the neighboring vertices across an edge over an iterator of query contexts.
    ///
    /// Each context in the `contexts` argument has an active vertex, which is
    /// either `None`, or a `Some(Self::Vertex)` value representing a vertex
    /// of type `type_name` defined in the schema.
    ///
    /// This function resolves the neighboring vertices for that active vertex.
    ///
    /// If the schema this adapter covers has no edges aside from starting edges,
    /// then this method will never be called and may be implemented as `unreachable!()`.
    ///
    /// The caller guarantees that:
    /// - `type_name` is a type or interface defined in the schema.
    /// - `edge_name` is an edge field on `type_name` defined in the schema.
    /// - Any parameters the edge requires per the schema have values provided.
    /// - When the active vertex is `Some(...)`, it's a vertex of type `type_name`:
    ///   either its type is exactly `type_name`, or `type_name` is an interface that
    ///   the vertex's type implements.
    ///
    /// The returned iterator must satisfy these properties:
    /// - Produce `(context, neighbors)` tuples with an iterator of neighbor vertices for that edge.
    /// - Produce contexts in the same order as the input `contexts` iterator produced them.
    /// - Each neighboring vertex is of the type specified for that edge in the schema.
    /// - When a context's active vertex is None, it has an empty neighbors iterator.
    fn resolve_neighbors(
        &mut self,
        contexts: ContextIterator<'vertex, Self::Vertex>,
        type_name: &str,
        edge_name: &str,
        parameters: Option<&EdgeParameters>,
    ) -> ContextOutcomeIterator<'vertex, Self::Vertex, VertexIterator<'vertex, Self::Vertex>>;

    /// Attempt to coerce vertices to a subtype, over an iterator of query contexts.
    ///
    /// In this example query, the starting vertices of type `Foo` are coerced to `Bar`:
    /// ```graphql
    /// query {
    ///     Foo {
    ///         ... on Bar {
    ///             abc @output
    ///         }
    ///     }
    /// }
    /// ```
    ///
    /// Each context in the `contexts` argument has an active vertex, which is
    /// either `None`, or a `Some(Self::Vertex)` value representing a vertex
    /// of type `type_name` defined in the schema.
    ///
    /// This function checks whether the active vertex is of the specified subtype.
    ///
    /// If this adapter's schema contains no subtyping, then no type coercions are possible:
    /// this method will never be called and may be implemented as `unreachable!()`.
    ///
    /// The caller guarantees that:
    /// - `type_name` is an interface defined in the schema.
    /// - `coerce_to_type` is a type or interface that implements `type_name` in the schema.
    /// - When the active vertex is `Some(...)`, it's a vertex of type `type_name`:
    ///   either its type is exactly `type_name`, or `type_name` is an interface that
    ///   the vertex's type implements.
    ///
    /// The returned iterator must satisfy these properties:
    /// - Produce `(context, can_coerce)` tuples showing if the coercion succeded for that context.
    /// - Produce contexts in the same order as the input `contexts` iterator produced them.
    /// - Each neighboring vertex is of the type specified for that edge in the schema.
    /// - When a context's active vertex is `None`, its coercion outcome is `false`.
    fn resolve_coercion(
        &mut self,
        contexts: ContextIterator<'vertex, Self::Vertex>,
        type_name: &str,
        coerce_to_type: &str,
    ) -> ContextOutcomeIterator<'vertex, Self::Vertex, bool>;
}

impl<'token, T> Adapter<'token> for T
where
    T: BasicAdapter<'token>,
{
    type DataToken = T::Vertex;

    fn get_starting_tokens(
        &mut self,
        edge: std::sync::Arc<str>,
        parameters: Option<std::sync::Arc<EdgeParameters>>,
        _query_hint: InterpretedQuery,
        _vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = Self::DataToken> + 'token> {
        <Self as BasicAdapter>::resolve_starting_vertices(
            self,
            edge.as_ref(),
            parameters.as_deref(),
        )
    }

    fn project_property(
        &mut self,
        contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: std::sync::Arc<str>,
        field_name: std::sync::Arc<str>,
        _query_hint: InterpretedQuery,
        _vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = (DataContext<Self::DataToken>, FieldValue)> + 'token> {
        <Self as BasicAdapter>::resolve_property(
            self,
            contexts,
            current_type_name.as_ref(),
            field_name.as_ref(),
        )
    }

    fn project_neighbors(
        &mut self,
        contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: std::sync::Arc<str>,
        edge_name: std::sync::Arc<str>,
        parameters: Option<std::sync::Arc<EdgeParameters>>,
        _query_hint: InterpretedQuery,
        _vertex_hint: Vid,
        _edge_hint: Eid,
    ) -> Box<
        dyn Iterator<
                Item = (
                    DataContext<Self::DataToken>,
                    Box<dyn Iterator<Item = Self::DataToken> + 'token>,
                ),
            > + 'token,
    > {
        <Self as BasicAdapter>::resolve_neighbors(
            self,
            contexts,
            current_type_name.as_ref(),
            edge_name.as_ref(),
            parameters.as_deref(),
        )
    }

    fn can_coerce_to_type(
        &mut self,
        contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: std::sync::Arc<str>,
        coerce_to_type_name: std::sync::Arc<str>,
        _query_hint: InterpretedQuery,
        _vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = (DataContext<Self::DataToken>, bool)> + 'token> {
        <Self as BasicAdapter>::resolve_coercion(
            self,
            contexts,
            current_type_name.as_ref(),
            coerce_to_type_name.as_ref(),
        )
    }
}
