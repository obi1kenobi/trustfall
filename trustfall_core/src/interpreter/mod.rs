use std::{collections::BTreeMap, fmt::Debug, sync::Arc};

use async_graphql_parser::types::Type;
use itertools::Itertools;
use serde::{Deserialize, Serialize};

use crate::{
    ir::{
        indexed::IndexedQuery, types::is_argument_type_valid, EdgeParameters, Eid, FieldRef,
        FieldValue, Vid,
    },
    util::BTreeMapTryInsertExt,
};

use self::error::QueryArgumentsError;

pub mod basic_adapter;
pub mod error;
pub mod execution;
mod filtering;
pub mod macros;
pub mod replay;
pub mod trace;

#[derive(Debug, Clone)]
pub struct DataContext<DataToken: Clone + Debug> {
    pub current_token: Option<DataToken>,
    tokens: BTreeMap<Vid, Option<DataToken>>,
    values: Vec<FieldValue>,
    suspended_tokens: Vec<Option<DataToken>>,
    folded_contexts: BTreeMap<Eid, Vec<DataContext<DataToken>>>,
    folded_values: BTreeMap<(Eid, Arc<str>), Option<ValueOrVec>>,
    piggyback: Option<Vec<DataContext<DataToken>>>,
    imported_tags: BTreeMap<FieldRef, FieldValue>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
enum ValueOrVec {
    Value(FieldValue),
    Vec(Vec<ValueOrVec>),
}

impl ValueOrVec {
    fn as_mut_vec(&mut self) -> Option<&mut Vec<ValueOrVec>> {
        match self {
            ValueOrVec::Value(_) => None,
            ValueOrVec::Vec(v) => Some(v),
        }
    }
}

impl From<ValueOrVec> for FieldValue {
    fn from(v: ValueOrVec) -> Self {
        match v {
            ValueOrVec::Value(value) => value,
            ValueOrVec::Vec(v) => v.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(bound = "DataToken: Serialize, for<'de2> DataToken: Deserialize<'de2>")]
struct SerializableContext<DataToken>
where
    DataToken: Clone + Debug + Serialize,
    for<'d> DataToken: Deserialize<'d>,
{
    current_token: Option<DataToken>,
    tokens: BTreeMap<Vid, Option<DataToken>>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    values: Vec<FieldValue>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    suspended_tokens: Vec<Option<DataToken>>,

    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    folded_contexts: BTreeMap<Eid, Vec<DataContext<DataToken>>>,

    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    folded_values: BTreeMap<(Eid, Arc<str>), Option<ValueOrVec>>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    piggyback: Option<Vec<DataContext<DataToken>>>,

    /// Tagged values imported from an ancestor component of the one currently being evaluated.
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    imported_tags: BTreeMap<FieldRef, FieldValue>,
}

impl<DataToken> From<SerializableContext<DataToken>> for DataContext<DataToken>
where
    DataToken: Clone + Debug + Serialize,
    for<'d> DataToken: Deserialize<'d>,
{
    fn from(context: SerializableContext<DataToken>) -> Self {
        Self {
            current_token: context.current_token,
            tokens: context.tokens,
            values: context.values,
            suspended_tokens: context.suspended_tokens,
            folded_contexts: context.folded_contexts,
            folded_values: context.folded_values,
            piggyback: context.piggyback,
            imported_tags: context.imported_tags,
        }
    }
}

impl<DataToken> From<DataContext<DataToken>> for SerializableContext<DataToken>
where
    DataToken: Clone + Debug + Serialize,
    for<'d> DataToken: Deserialize<'d>,
{
    fn from(context: DataContext<DataToken>) -> Self {
        Self {
            current_token: context.current_token,
            tokens: context.tokens,
            values: context.values,
            suspended_tokens: context.suspended_tokens,
            folded_contexts: context.folded_contexts,
            folded_values: context.folded_values,
            piggyback: context.piggyback,
            imported_tags: context.imported_tags,
        }
    }
}

impl<DataToken: Clone + Debug> DataContext<DataToken> {
    pub fn new(token: Option<DataToken>) -> DataContext<DataToken> {
        DataContext {
            current_token: token,
            piggyback: None,
            tokens: Default::default(),
            values: Default::default(),
            suspended_tokens: Default::default(),
            folded_contexts: Default::default(),
            folded_values: Default::default(),
            imported_tags: Default::default(),
        }
    }

    fn record_token(&mut self, vid: Vid) {
        self.tokens
            .insert_or_error(vid, self.current_token.clone())
            .unwrap();
    }

    fn activate_token(self, vid: &Vid) -> DataContext<DataToken> {
        DataContext {
            current_token: self.tokens[vid].clone(),
            tokens: self.tokens,
            values: self.values,
            suspended_tokens: self.suspended_tokens,
            folded_contexts: self.folded_contexts,
            folded_values: self.folded_values,
            piggyback: self.piggyback,
            imported_tags: self.imported_tags,
        }
    }

    fn split_and_move_to_token(&self, new_token: Option<DataToken>) -> DataContext<DataToken> {
        DataContext {
            current_token: new_token,
            tokens: self.tokens.clone(),
            values: self.values.clone(),
            suspended_tokens: self.suspended_tokens.clone(),
            folded_contexts: self.folded_contexts.clone(),
            folded_values: self.folded_values.clone(),
            piggyback: None,
            imported_tags: self.imported_tags.clone(),
        }
    }

    fn move_to_token(self, new_token: Option<DataToken>) -> DataContext<DataToken> {
        DataContext {
            current_token: new_token,
            tokens: self.tokens,
            values: self.values,
            suspended_tokens: self.suspended_tokens,
            folded_contexts: self.folded_contexts,
            folded_values: self.folded_values,
            piggyback: self.piggyback,
            imported_tags: self.imported_tags,
        }
    }

    fn ensure_suspended(mut self) -> DataContext<DataToken> {
        if let Some(token) = self.current_token {
            self.suspended_tokens.push(Some(token));
            DataContext {
                current_token: None,
                tokens: self.tokens,
                values: self.values,
                suspended_tokens: self.suspended_tokens,
                folded_contexts: self.folded_contexts,
                folded_values: self.folded_values,
                piggyback: self.piggyback,
                imported_tags: self.imported_tags,
            }
        } else {
            self
        }
    }

    fn ensure_unsuspended(mut self) -> DataContext<DataToken> {
        match self.current_token {
            None => {
                let current_token = self.suspended_tokens.pop().unwrap();
                DataContext {
                    current_token,
                    tokens: self.tokens,
                    values: self.values,
                    suspended_tokens: self.suspended_tokens,
                    folded_contexts: self.folded_contexts,
                    folded_values: self.folded_values,
                    piggyback: self.piggyback,
                    imported_tags: self.imported_tags,
                }
            }
            Some(_) => self,
        }
    }
}

impl<DataToken: Debug + Clone + PartialEq> PartialEq for DataContext<DataToken> {
    fn eq(&self, other: &Self) -> bool {
        self.current_token == other.current_token
            && self.tokens == other.tokens
            && self.values == other.values
            && self.suspended_tokens == other.suspended_tokens
            && self.folded_contexts == other.folded_contexts
            && self.piggyback == other.piggyback
            && self.imported_tags == other.imported_tags
    }
}

impl<DataToken: Debug + Clone + PartialEq + Eq> Eq for DataContext<DataToken> {}

impl<DataToken> Serialize for DataContext<DataToken>
where
    DataToken: Debug + Clone + Serialize,
    for<'d> DataToken: Deserialize<'d>,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        // TODO: eventually maybe write a proper (de)serialize?
        SerializableContext::from(self.clone()).serialize(serializer)
    }
}

impl<'de, DataToken> Deserialize<'de> for DataContext<DataToken>
where
    DataToken: Debug + Clone + Serialize,
    for<'d> DataToken: Deserialize<'d>,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        // TODO: eventually maybe write a proper (de)serialize?
        SerializableContext::deserialize(deserializer).map(DataContext::from)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InterpretedQuery {
    pub indexed_query: Arc<IndexedQuery>,
    pub arguments: Arc<BTreeMap<Arc<str>, FieldValue>>,
}

impl InterpretedQuery {
    #[inline]
    pub fn from_query_and_arguments(
        indexed_query: Arc<IndexedQuery>,
        arguments: Arc<BTreeMap<Arc<str>, FieldValue>>,
    ) -> Result<Self, QueryArgumentsError> {
        let mut errors = vec![];

        let mut missing_arguments = vec![];
        for (variable_name, variable_type) in &indexed_query.ir_query.variables {
            match arguments.get(variable_name) {
                Some(argument_value) => {
                    // Ensure the provided argument value is valid for the variable's inferred type.
                    if let Err(e) = validate_argument_type(
                        variable_name.as_ref(),
                        variable_type,
                        argument_value,
                    ) {
                        errors.push(e);
                    }
                }
                None => {
                    missing_arguments.push(variable_name.as_ref());
                }
            }
        }
        if !missing_arguments.is_empty() {
            errors.push(QueryArgumentsError::MissingArguments(
                missing_arguments
                    .into_iter()
                    .map(|x| x.to_string())
                    .collect(),
            ));
        }

        let unused_arguments = arguments
            .keys()
            .map(|x| x.as_ref())
            .filter(|arg| !indexed_query.ir_query.variables.contains_key(*arg))
            .collect_vec();
        if !unused_arguments.is_empty() {
            errors.push(QueryArgumentsError::UnusedArguments(
                unused_arguments
                    .into_iter()
                    .map(|x| x.to_string())
                    .collect(),
            ));
        }

        if errors.is_empty() {
            Ok(Self {
                indexed_query,
                arguments,
            })
        } else {
            Err(errors.into())
        }
    }
}

fn validate_argument_type(
    variable_name: &str,
    variable_type: &Type,
    argument_value: &FieldValue,
) -> Result<(), QueryArgumentsError> {
    if is_argument_type_valid(variable_type, argument_value) {
        Ok(())
    } else {
        Err(QueryArgumentsError::ArgumentTypeError(
            variable_name.to_string(),
            variable_type.to_string(),
            argument_value.to_owned(),
        ))
    }
}

pub trait Adapter<'token> {
    type DataToken: Clone + Debug + 'token;

    fn get_starting_tokens(
        &mut self,
        edge: Arc<str>,
        parameters: Option<Arc<EdgeParameters>>,
        query_hint: InterpretedQuery,
        vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = Self::DataToken> + 'token>;

    #[allow(clippy::type_complexity)]
    fn project_property(
        &mut self,
        data_contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: Arc<str>,
        field_name: Arc<str>,
        query_hint: InterpretedQuery,
        vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = (DataContext<Self::DataToken>, FieldValue)> + 'token>;

    #[allow(clippy::type_complexity)]
    #[allow(clippy::too_many_arguments)]
    fn project_neighbors(
        &mut self,
        data_contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: Arc<str>,
        edge_name: Arc<str>,
        parameters: Option<Arc<EdgeParameters>>,
        query_hint: InterpretedQuery,
        vertex_hint: Vid,
        edge_hint: Eid,
    ) -> Box<
        dyn Iterator<
                Item = (
                    DataContext<Self::DataToken>,
                    Box<dyn Iterator<Item = Self::DataToken> + 'token>,
                ),
            > + 'token,
    >;

    fn can_coerce_to_type(
        &mut self,
        data_contexts: Box<dyn Iterator<Item = DataContext<Self::DataToken>> + 'token>,
        current_type_name: Arc<str>,
        coerce_to_type_name: Arc<str>,
        query_hint: InterpretedQuery,
        vertex_hint: Vid,
    ) -> Box<dyn Iterator<Item = (DataContext<Self::DataToken>, bool)> + 'token>;
}
