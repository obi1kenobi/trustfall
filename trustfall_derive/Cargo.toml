[package]
name = "trustfall_derive"
version = "0.3.1"
license = "Apache-2.0"
description = "Derive macros for the trustfall query engine."
repository = "https://github.com/obi1kenobi/trustfall"
readme = "../README.md"
edition.workspace = true
rust-version.workspace = true
authors.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
quote = { workspace = true }
syn = { workspace = true }
proc-macro2 = { workspace = true }

[dev-dependencies]
trybuild = "1.0"
trustfall = { path = "../trustfall" }
