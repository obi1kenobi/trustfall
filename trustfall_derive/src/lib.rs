use quote::quote;
use syn::punctuated::Punctuated;

const TRUSTFALL_ATTRIBUTE: &str = "trustfall";
const SKIP_CONVERSION_ATTRIBUTE: &str = "skip_conversion";

/// Adds `typename()` and `as_<variant>()` methods on an enum being used as a Trustfall vertex.
///
/// The `typename()` method is part of Trustfall's [`Typename`] trait.
///
/// For example:
/// ```rust
/// # use trustfall_derive::TrustfallEnumVertex;
/// #
/// #[derive(Debug, Clone, TrustfallEnumVertex)]
/// enum Vertex {
///     User(String),
///     Message { author: String, content: String },
///     EmptyVariant,
/// }
/// ```
/// will get the following implementations:
/// ```rust
/// # use trustfall_core::interpreter::Typename;
/// #
/// # #[derive(Debug, Clone)]
/// # enum Vertex {
/// #     User(String),
/// #     Message { author: String, content: String },
/// #     EmptyVariant,
/// # }
/// #
/// impl Typename for Vertex {
///     fn typename(&self) -> &'static str {
///         match self {
///             Self::User { .. } => "User",
///             Self::Message { .. } => "Message",
///             Self::EmptyVariant { .. } => "EmptyVariant",
///         }
///     }
/// }
///
/// impl Vertex {
///     fn as_user(&self) -> Option<&String> {
///         match self {
///             Self::User(x) => Some(x),
///             _ => None,
///         }
///     }
///
///     fn as_message(&self) -> Option<(&String, &String)> {
///         match self {
///             Self::Message { author, content } => Some((author, content)),
///             _ => None,
///         }
///     }
///
///     fn as_empty_variant(&self) -> Option<()> {
///         match self {
///             Self::EmptyVariant => Some(()),
///             _ => None,
///         }
///     }
/// }
/// ```
///
/// [`Typename`](trustfall_core::interpreter::Typename)
#[proc_macro_derive(TrustfallEnumVertex, attributes(trustfall))]
pub fn trustfall_enum_vertex_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    match syn::parse(input) {
        Ok(ast) => impl_trustfall_enum_vertex(&ast).unwrap_or_else(syn::Error::into_compile_error),
        Err(e) => e.into_compile_error(),
    }
    .into()
}

fn impl_trustfall_enum_vertex(ast: &syn::DeriveInput) -> syn::Result<proc_macro2::TokenStream> {
    let name = &ast.ident;
    let (impl_generics, ty_generics, where_clause) = ast.generics.split_for_impl();

    let variants = match &ast.data {
        syn::Data::Enum(d) => &d.variants,
        _ => {
            return Err(syn::Error::new_spanned(
                ast,
                "only enums can derive TrustfallEnumVertex",
            ))
        }
    };

    let arms = variants
        .iter()
        .map(generate_typename_arm)
        .reduce(|mut acc, e| {
            acc.extend(e);
            acc
        })
        .unwrap_or_default();

    let conversions = variants
        .iter()
        .map(generate_conversion_method)
        .collect::<Result<Vec<_>, _>>()?
        .into_iter()
        .reduce(|mut acc, e| {
            acc.extend(e);
            acc
        });

    let conversions_impl = if let Some(conversions) = conversions {
        quote! {
            #[automatically_derived]
            impl #impl_generics #name #ty_generics #where_clause {
                #conversions
            }
        }
    } else {
        Default::default()
    };

    let gen = quote! {
        #[automatically_derived]
        impl #impl_generics ::trustfall_core::interpreter::Typename for #name #ty_generics #where_clause {
            fn typename(&self) -> &'static str {
                match self {
                    #arms

                    #[allow(unreachable_code)]
                    _ => unreachable!("this arm exists only for uninhabited enums"),
                }
            }
        }
        #conversions_impl
    };
    Ok(gen)
}

fn generate_typename_arm(variant: &syn::Variant) -> proc_macro2::TokenStream {
    let variant_ident = &variant.ident;
    let variant_name = variant_ident.to_string();
    let typename = proc_macro2::Literal::string(&variant_name);
    syn::parse_quote! {
        Self::#variant_ident { .. } => #typename,
    }
}

fn generate_conversion_method(variant: &syn::Variant) -> syn::Result<proc_macro2::TokenStream> {
    // Check if we should skip generating the conversion method
    // because of a `#[trustfall(skip_conversion)]` attribute on the variant.
    for attr in &variant.attrs {
        if let Some(ident) = attr.path.get_ident() {
            if ident != TRUSTFALL_ATTRIBUTE {
                // Not one of our attributes, skip.
                continue;
            }
        }

        // If we ever add more attribute contents, here's how to make the parsing smarter:
        // https://blog.turbo.fish/proc-macro-parsing/
        match attr.parse_meta()? {
            syn::Meta::Path { .. } => {
                return Err(syn::Error::new_spanned(
                    attr,
                    "no arguments found, did you mean `#[trustfall(skip_conversion)]`?",
                ));
            }
            syn::Meta::List(values) => {
                let mut skipping = false;
                for nested in values.nested.iter() {
                    match nested {
                        syn::NestedMeta::Meta(syn::Meta::Path(path)) => {
                            if let Some(ident) = path.get_ident() {
                                if ident == SKIP_CONVERSION_ATTRIBUTE {
                                    skipping = true;
                                } else {
                                    return Err(syn::Error::new_spanned(
                                        nested,
                                        "unexpected arguments found, did you mean `#[trustfall(skip_conversion)]`?",
                                    ));
                                }
                            }
                        }
                        _ => {
                            return Err(syn::Error::new_spanned(
                                attr,
                                "unexpected arguments found, did you mean `#[trustfall(skip_conversion)]`?",
                            ));
                        }
                    }
                }
                if skipping {
                    return Ok(Default::default());
                }
            }
            syn::Meta::NameValue(name_value) => {
                return Err(syn::Error::new_spanned(
                    &name_value.lit,
                    "unexpected arguments found, did you mean `#[trustfall(skip_conversion)]`?",
                ));
            }
        };
    }

    let variant_ident = &variant.ident;
    let variant_name = variant_ident.to_string();
    let conversion_name = syn::Ident::new(
        &format!("as_{}", to_lower_snake_case(&variant_name)),
        proc_macro2::Span::call_site(),
    );

    match &variant.fields {
        syn::Fields::Named(named_fields) if !named_fields.named.is_empty() => {
            if named_fields.named.len() == 1 {
                // Struct variants with only a single field return `Option<&ThatField>`.
                let named_arg = &named_fields.named[0];
                let field_name = named_arg
                    .ident
                    .as_ref()
                    .expect("struct variant field had no name");
                let field_type = &named_arg.ty;
                Ok(syn::parse_quote! {
                    pub(crate) fn #conversion_name(&self) -> Option<&#field_type> {
                        match self {
                            Self::#variant_ident { #field_name } => Some(#field_name),
                            _ => None,
                        }
                    }
                })
            } else {
                // Struct variants with multiple fields return
                // `Option<(&FirstField, &SecondField, ...)>`
                // in the order the fields were defined.
                let final_type: proc_macro2::TokenStream =
                    tuple_of_field_types(&named_fields.named);

                let mut fields = syn::punctuated::Punctuated::<_, syn::Token![,]>::new();
                for field in named_fields.named.iter() {
                    let field_name = field
                        .ident
                        .as_ref()
                        .expect("struct variant field had no name");
                    fields.push(field_name);
                }
                Ok(syn::parse_quote! {
                    pub(crate) fn #conversion_name(&self) -> Option<#final_type> {
                        match self {
                            Self::#variant_ident { #fields } => Some((#fields)),
                            _ => None,
                        }
                    }
                })
            }
        }
        syn::Fields::Unnamed(tuple_fields) if !tuple_fields.unnamed.is_empty() => {
            if tuple_fields.unnamed.len() == 1 {
                // Tuple variants with only a single field return `Option<&ThatField>`.
                let field_type = &tuple_fields.unnamed[0].ty;
                Ok(syn::parse_quote! {
                    pub(crate) fn #conversion_name(&self) -> Option<&#field_type> {
                        match self {
                            Self::#variant_ident(x) => Some(x),
                            _ => None,
                        }
                    }
                })
            } else {
                // Tuple variants with multiple fields return
                // `Option<(&FirstField, &SecondField, ...)>`.
                let final_type: proc_macro2::TokenStream =
                    tuple_of_field_types(&tuple_fields.unnamed);
                let mut fields = syn::punctuated::Punctuated::<_, syn::Token![,]>::new();
                for (i, _) in tuple_fields.unnamed.iter().enumerate() {
                    fields.push(quote::format_ident!("x{i}"));
                }
                Ok(syn::parse_quote! {
                    pub(crate) fn #conversion_name(&self) -> Option<#final_type> {
                        match self {
                            Self::#variant_ident(#fields) => Some((#fields)),
                            _ => None,
                        }
                    }
                })
            }
        }
        _ => {
            // Either unit variant, or fieldless struct/tuple variant.
            Ok(syn::parse_quote! {
                pub(crate) fn #conversion_name(&self) -> Option<()> {
                    match self {
                        Self::#variant_ident => Some(()),
                        _ => None,
                    }
                }
            })
        }
    }
}

/// Returns a tuple of references to all field types.
/// The input must contain more than one field.
fn tuple_of_field_types(
    fields: &Punctuated<syn::Field, syn::Token![,]>,
) -> proc_macro2::TokenStream {
    if fields.len() > 1 {
        let mut punct = syn::punctuated::Punctuated::<_, syn::Token![,]>::new();
        for field in fields.iter() {
            let ty = &field.ty;
            punct.push(quote!(&#ty))
        }
        quote!((#punct))
    } else {
        panic!(
            "list of fields had {} field(s), which is not more than one field",
            fields.len()
        );
    }
}

fn to_lower_snake_case(value: &str) -> String {
    let mut result = String::with_capacity(value.len());
    let mut last = '_';
    for c in value.chars() {
        if c.is_uppercase() {
            if last != '_' {
                result.push('_');
            }
            result.extend(c.to_lowercase());
        } else {
            result.push(c);
        }
        last = c;
    }
    result
}
