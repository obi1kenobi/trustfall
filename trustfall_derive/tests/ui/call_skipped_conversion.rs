use trustfall_core::interpreter::Typename;
use trustfall_derive::TrustfallEnumVertex;

#[derive(Debug, Clone, TrustfallEnumVertex)]
enum TwoVariants {
    #[trustfall(skip_conversion)]
    First,
    Second,
}

fn main() {
    let first = TwoVariants::First;
    assert_eq!("First", first.typename());
    assert_eq!(None, first.as_second());

    // this method should not exist, expecting an error here
    first.as_first();

    let second = TwoVariants::Second;
    assert_eq!("Second", second.typename());
    assert_eq!(Some(()), second.as_second());

    // this method should not exist, expecting an error here
    second.as_first();
}
