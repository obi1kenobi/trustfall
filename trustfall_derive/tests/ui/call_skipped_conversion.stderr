error[E0599]: no method named `as_first` found for enum `TwoVariants` in the current scope
  --> tests/ui/call_skipped_conversion.rs:17:11
   |
5  | enum TwoVariants {
   | ---------------- method `as_first` not found for this enum
...
17 |     first.as_first();
   |           ^^^^^^^^ method not found in `TwoVariants`

error[E0599]: no method named `as_first` found for enum `TwoVariants` in the current scope
  --> tests/ui/call_skipped_conversion.rs:24:12
   |
5  | enum TwoVariants {
   | ---------------- method `as_first` not found for this enum
...
24 |     second.as_first();
   |            ^^^^^^^^ method not found in `TwoVariants`
